src/rgw/services/svc_bucket_sync_sobj.cc

```diff
@@ -0,0 +1,872 @@
+    }
+
+    bool empty() const {
+      return instances.empty();
+    }
+
+    void clear() {
+      instances.clear();
+    }
+
+    void get_entities(const rgw_bucket& bucket,
+                      std::set<rgw_bucket> *result) const {
+      auto iter = instances.find(bucket);
+      if (iter == instances.end()) {
+        return;
+      }
+      iter->second.get_entities(result);
+    }
+  } info;
+
+  RGWSI_BS_SObj_HintIndexObj(RGWSI_SysObj *_sysobj_svc,
+                             const rgw_raw_obj& _obj) : cct(_sysobj_svc->ctx()),
+                                                        obj_ctx(_sysobj_svc->init_obj_ctx()),
+                                                        obj(_obj),
+                                                        sysobj(obj_ctx.get_obj(obj))
+  {
+    svc.sysobj = _sysobj_svc;
+  }
+
+  template <typename C1, typename C2>
+  int update(const rgw_bucket& entity,
+             const RGWBucketInfo& info_source,
+             C1 *add,
+             C2 *remove,
+             optional_yield y);
+
+private:
+  template <typename C1, typename C2>
+  void update_entries(const rgw_bucket& info_source,
+                      const obj_version& info_source_ver,
+                      C1 *add,
+                      C2 *remove,
+                      single_instance_info *instance);
+
+  int read(optional_yield y);
+  int flush(optional_yield y);
+
+  void invalidate() {
+    has_data = false;
+    info.clear();
+  }
+
+  void get_entities(const rgw_bucket& bucket,
+                    std::set<rgw_bucket> *result) const {
+    info.get_entities(bucket, result);
+  }
+};
+WRITE_CLASS_ENCODER(RGWSI_BS_SObj_HintIndexObj::bi_entry)
+WRITE_CLASS_ENCODER(RGWSI_BS_SObj_HintIndexObj::single_instance_info)
+WRITE_CLASS_ENCODER(RGWSI_BS_SObj_HintIndexObj::info_map)
+
+template <typename C1, typename C2>
+int RGWSI_BS_SObj_HintIndexObj::update(const rgw_bucket& entity,
+                                       const RGWBucketInfo& info_source,
+                                       C1 *add,
+                                       C2 *remove,
+                                       optional_yield y)
+{
+  int r = 0;
+
+  auto& info_source_ver = info_source.objv_tracker.read_version;
+
+#define MAX_RETRIES 25
+
+  for (int i = 0; i < MAX_RETRIES; ++i) {
+    if (!has_data) {
+      r = read(y);
+      if (r < 0) {
+        ldout(cct, 0) << "ERROR: cannot update hint index: failed to read: r=" << r << dendl;
+        return r;
+      }
+    }
+
+    auto& instance = info.instances[entity];
+
+    update_entries(info_source.bucket,
+                   info_source_ver,
+                   add, remove,
+                   &instance);
+
+    if (instance.empty()) {
+      info.instances.erase(entity);
+    }
+
+    r = flush(y);
+    if (r >= 0) {
+      return 0;
+    }
+
+    if (r != -ECANCELED) {
+      ldout(cct, 0) << "ERROR: failed to flush hint index: obj=" << obj << " r=" << r << dendl;
+      return r;
+    }
```

Comment:
@cbodley no need to call `invalidate()` as `read()` resets the version tracker. Can remove `invalidate()` if it's unused.

Reply:
it won't call `read()` again though, because has_data will be true for the next iteration of the loop. i think `invalidate()` is needed to clear `has_data`

Reply:
@cbodley ack: https://github.com/ceph/ceph/pull/31686/commits/ce962f3470a510d598f9c8bc0b7248cbf1c37cdc